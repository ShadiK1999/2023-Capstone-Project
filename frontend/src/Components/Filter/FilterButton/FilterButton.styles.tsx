import { style } from 'typestyle';

export const container = style({
  position: 'relative',
});

export const filterButtonContent = style({
  width: '150px',
});
