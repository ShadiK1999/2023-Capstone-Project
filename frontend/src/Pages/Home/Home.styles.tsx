import { style } from 'typestyle';

export const Container = style({
  width: '100%',
  display: 'grid',
  marginTop: '75px',
  columnGap: '8px',
  rowGap: '30px',
  gridTemplateColumns: '1fr 400px',
  gridTemplateRows: '53px minmax(0, 1fr)',
});
