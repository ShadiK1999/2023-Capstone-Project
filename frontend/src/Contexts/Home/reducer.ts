import { HomeState, Filter } from './state';
import { IsochroneRingsOutput } from '../../Services/Isochrone';
import { produce } from 'immer';
import { DisplayPoint } from '../../Models/DeliveryPoint';

export type HomeAction =
  | {
      type: 'ISOCHRONE_READY';
      payload: IsochroneRingsOutput;
    }
  | {
      type: 'ADDRESS_CHANGED';
      payload: google.maps.LatLngLiteral;
    }
  | {
      type: 'FETCH_DELIVERY_POINTS_SUCCESS';
      payload: DisplayPoint[];
    }
  | {
      type: 'START_LOADING';
    }
  | {
      type: 'LOAD_SUCCESSFUL';
    }
  | {
      type: 'FILTER_INPUT_CHANGED';
      payload: Filter;
    };

export const homeReducer = produce((draft: HomeState, action: HomeAction) => {
  switch (action.type) {
    case 'ISOCHRONE_READY':
      draft.map.isochroneRings = action.payload;
      break;
    case 'ADDRESS_CHANGED':
      draft.map.currentLocation = action.payload;
      break;
    case 'FILTER_INPUT_CHANGED':
      draft.filter = action.payload;
      break;
    case 'FETCH_DELIVERY_POINTS_SUCCESS':
      draft.deliveryPoints = action.payload;
      break;
    case 'START_LOADING':
      draft.isLoading = true;
      break;
    case 'LOAD_SUCCESSFUL':
      draft.isLoading = false;
      break;
  }
});
