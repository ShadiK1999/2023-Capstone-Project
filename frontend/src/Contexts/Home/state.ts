import { Dispatch, createContext } from 'react';
import { HomeAction } from './reducer';
import { DisplayPoint } from '../../Models/DeliveryPoint';

type IsochroneRings = {
  inner: google.maps.LatLngLiteral[];
  middle: google.maps.LatLngLiteral[];
  outer: google.maps.LatLngLiteral[];
};

type Map = {
  currentLocation: google.maps.LatLngLiteral;
  isochroneRings: IsochroneRings | null;
};

export type DayOfWeekIndex = 0 | 1 | 2 | 3 | 4 | 5 | 6;

export type Filter = {
  type: 'distance' | 'duration';
  mode: 'WALKING' | 'DRIVING' | 'CYCLING';
  value: number;
  dayOfWeek: DayOfWeekIndex;
  startHour: number;
  startMin: number;
  endHour: number;
  endMin: number;
};

export type HomeState = {
  map: Map;
  deliveryPoints: DisplayPoint[] | null;
  isLoading: boolean;
  filter: Filter;
};

export const initialState: HomeState = {
  map: {
    currentLocation: {
      lat: -37.8124,
      lng: 144.9623,
    },
    isochroneRings: null,
  },
  deliveryPoints: null,
  isLoading: false,
  filter: {
    type: 'duration',
    mode: 'WALKING',
    value: 5,
    dayOfWeek: 0,
    startHour: 9,
    startMin: 0,
    endHour: 17,
    endMin: 0,
  },
};

export const HomeContext = createContext<HomeState>(initialState);
export const HomeDispatchContext = createContext<Dispatch<HomeAction>>(() => void 0);
