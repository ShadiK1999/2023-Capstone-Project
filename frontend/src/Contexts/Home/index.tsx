import { FC, ReactNode, useReducer } from 'react';
import { HomeContext, HomeDispatchContext, initialState } from './state';
import { homeReducer } from './reducer';

type ComponentProps = {
  children: ReactNode;
};

const HomeProvider: FC<ComponentProps> = ({ children }) => {
  const [state, dispatch] = useReducer(homeReducer, initialState);

  return (
    <HomeContext.Provider value={state}>
      <HomeDispatchContext.Provider value={dispatch}>{children}</HomeDispatchContext.Provider>
    </HomeContext.Provider>
  );
};

export { HomeProvider };
