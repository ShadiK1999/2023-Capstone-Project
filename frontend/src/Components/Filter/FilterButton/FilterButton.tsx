import { useContext, useState } from 'react';
import { HomeContext } from '../../../Contexts/Home/state';
import { PlusMinusButton } from '../PlusMinusButton';
import { Modal } from '../../Modal';
import { FilterTravelMode, FilterTravelTime, FilterDayOfWeek, FilterTimeOfDay } from '../Filters';
import * as style from './FilterButton.styles';

export const FilterButton: React.FC = () => {
  const {
    filter: { mode, value },
  } = useContext(HomeContext);

  const [showModal, setShowModal] = useState(false);

  return (
    <div className={style.container}>
      <PlusMinusButton
        leftButton={{
          icon: 'FILTER',
          onClick: () => setShowModal(true),
        }}
        rightButton={{
          onClick: () => setShowModal(true),
        }}
      >
        <div className={style.filterButtonContent}>{mode.toLowerCase() + ' ' + value + ' min'}</div>
      </PlusMinusButton>
      <Modal title={'Filter'} isOpen={showModal} onClose={() => setShowModal(false)}>
        <FilterTravelMode />
        <FilterTravelTime />
        <FilterDayOfWeek />
        <FilterTimeOfDay />
      </Modal>
    </div>
  );
};
