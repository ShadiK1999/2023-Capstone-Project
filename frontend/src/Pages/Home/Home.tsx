import { Map, InputBar, SideBar } from '../../Components';
import { HomeProvider } from '../../Contexts';

import * as styles from './Home.styles';

const Home = () => {
  return (
    <HomeProvider>
      <PageInternal />
    </HomeProvider>
  );
};

const PageInternal = () => {
  return (
    <div className={styles.Container}>
      <InputBar />
      <div></div>
      <Map />
      <SideBar />
    </div>
  );
};

export { Home };
