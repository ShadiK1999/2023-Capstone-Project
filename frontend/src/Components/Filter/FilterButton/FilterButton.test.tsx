import { render, fireEvent } from '@testing-library/react';
import { FilterButton } from './FilterButton';
import { describe, it, expect, beforeEach, afterEach } from 'vitest';
import { HomeContext, Filter, HomeState, initialState } from '../../../Contexts/Home/state';

const mockedFilter: Filter = {
  type: 'duration',
  mode: 'WALKING',
  value: 15,
  dayOfWeek: 0,
  startHour: 8,
  startMin: 0,
  endHour: 8,
  endMin: 30,
};

const mockedState: HomeState = {
  ...initialState,
  filter: mockedFilter,
};

describe('Matches DOM Snapshot', () => {
  beforeEach(() => {
    const modalRoot = document.createElement('div');
    modalRoot.setAttribute('id', 'modal-root');
    document.body.appendChild(modalRoot);
  });

  afterEach(() => {
    const modalRoot = document.getElementById('modal-root');
    if (modalRoot) {
      document.body.removeChild(modalRoot);
    }
  });

  it('Matches DOM Snapshot', () => {
    const tree = render(<FilterButton />);
    expect(tree).toMatchSnapshot();
  });

  it('Has correct icons', () => {
    const { getByAltText } = render(<FilterButton />);
    expect(getByAltText('FILTER')).toBeInTheDocument();
    expect(getByAltText('PLUS')).toBeInTheDocument();
  });

  it('Shows filter panel when plus button clicked', () => {
    const { getByAltText, queryByText, getByText } = render(
      <HomeContext.Provider value={mockedState}>
        <FilterButton />
      </HomeContext.Provider>,
    );
    expect(queryByText('Travel Time')).not.toBeInTheDocument();

    const plusButton = getByAltText('PLUS');
    fireEvent.click(plusButton);

    // Travel time is an arbitrary string that is only present in the filter panel.
    expect(queryByText('Travel Time')).toBeInTheDocument();

    const saveButton = getByText('Close');
    fireEvent.click(saveButton);

    expect(queryByText('Travel Time')).not.toBeInTheDocument();
  });
});
