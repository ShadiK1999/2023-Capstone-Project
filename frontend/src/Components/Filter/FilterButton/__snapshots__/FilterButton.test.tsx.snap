// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Matches DOM Snapshot > Matches DOM Snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div
      id="modal-root"
    />
    <div>
      <div
        class="f1mgyl9u"
      >
        <div
          class="f723edx"
        >
          <div
            class="f2ey5g1"
          >
            <img
              alt="FILTER"
              src="/src/Components/Filter/PlusMinusButton/Assets/Filter.svg"
            />
          </div>
          <div
            class="f1sz7xol"
          >
            <div
              class="f1vjmc79"
            >
              walking 5 min
            </div>
          </div>
          <div
            class="f1fozamg"
          >
            <img
              alt="PLUS"
              src="/src/Components/Filter/PlusMinusButton/Assets/Plus.svg"
            />
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="f1mgyl9u"
    >
      <div
        class="f723edx"
      >
        <div
          class="f2ey5g1"
        >
          <img
            alt="FILTER"
            src="/src/Components/Filter/PlusMinusButton/Assets/Filter.svg"
          />
        </div>
        <div
          class="f1sz7xol"
        >
          <div
            class="f1vjmc79"
          >
            walking 5 min
          </div>
        </div>
        <div
          class="f1fozamg"
        >
          <img
            alt="PLUS"
            src="/src/Components/Filter/PlusMinusButton/Assets/Plus.svg"
          />
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
